File: WOT-iOS/WOT-iOS/ui/WOTTankPivotViewController/WOTTankPivotViewController.m
//
//  WOTTankPivotViewController.m
//  WOT-iOS
//
//  Created by Pavel Yeshchyk on 8/12/15.
//  Copyright (c) 2015 Pavel Yeshchyk. All rights reserved.
//

#import "WOTTankPivotViewController.h"
#import "WOTTankPivotDataCollectionViewCell.h"
#import "WOTTankPivotFilterCollectionViewCell.h"
#import "WOTTankPivotFixedCollectionViewCell.h"
#import "WOTTankPivotEmptyCollectionViewCell.h"
#import "Tanks+DPM.h"
#import "WOTTankListSettingsDatasource.h"
#import "WOTNode+PivotFactory.h"

@interface WOTTankPivotViewController(NodeCreator) <WOTNodeCreatorProtocol>
@end

@implementation WOTTankPivotViewController(NodeCreator)

- (id<WOTNodeProtocol> _Nonnull)createNodeWithName:(NSString * _Nonnull)name {
    id<WOTNodeProtocol> result = [[WOTNode alloc] initWithName: name];
    result.isVisible = false;
    return result;
}

- (id<WOTNodeProtocol> _Nonnull)createNodeWithFetchedObject:(id<NSFetchRequestResult> _Nullable)fetchedObject byPredicate:(NSPredicate * _Nullable)byPredicate {
    return [WOTNodeFactory pivotDataNodeForPredicate:byPredicate andTanksObject:fetchedObject];
}

@end

@interface WOTTankPivotViewController(WOTDataFetchControllerDelegateProtocol) <WOTDataFetchControllerDelegateProtocol>
@end

@implementation WOTTankPivotViewController(WOTDataFetchControllerDelegateProtocol)
@dynamic fetchRequest;

- (NSFetchRequest *)fetchRequest {
    NSFetchRequest * result = [[NSFetchRequest alloc] initWithEntityName:@"Tanks"];
    result.sortDescriptors = [self sortDescriptors];
    result.predicate = [self fetchCustomPredicate];
    return result;
}

- (NSArray *) sortDescriptors {
    NSMutableArray *result = [[[WOTTankListSettingsDatasource sharedInstance] sortBy] mutableCopy];
    NSSortDescriptor *descriptor = [[NSSortDescriptor alloc] initWithKey:@"tank_id" ascending:YES];
    [result addObject:descriptor];
    return result;
}

- (NSPredicate *) fetchCustomPredicate {
    NSPredicate *level6 = [NSPredicate predicateWithFormat:@"level != %d", 600];
//    NSPredicate *level7 = [NSPredicate predicateWithFormat:@"level == %d", 7];
    return [NSCompoundPredicate orPredicateWithSubpredicates:@[level6]];
}

@end

@interface WOTTankPivotViewController () <UICollectionViewDataSource, WOTDataModelListener>

@property (nonatomic, weak)IBOutlet UICollectionView *collectionView;
@property (nonatomic, weak)IBOutlet WOTTankPivotLayout *flowLayout;

@property (nonatomic, strong) NSArray *fixedRowsTopLevel;
@property (nonatomic, strong) NSDictionary *fixedRowsChildren;

@property (nonatomic, strong) NSFetchedResultsController *fetchedResultController;
@property (nonatomic, readonly) NSArray *sortDescriptors;

@property (nonatomic, strong) WOTPivotDataModel *model;
@property (nonatomic, strong) id<WOTDataFetchControllerProtocol> fetchController;


@end

@implementation WOTTankPivotViewController

- (void)viewDidLoad {
    
    [super viewDidLoad];

    [self.flowLayout setRelativeContentSizeBlock:^CGSize{
        return self.model.contentSize;
    }];

    [self.flowLayout setItemRelativeRectCallback:^CGRect(NSIndexPath *indexPath) {
        return [self.model itemRectAtIndexPath:indexPath];
    }];

    [self.flowLayout setItemLayoutStickyType:^PivotStickyType(NSIndexPath *indexPath) {
        id<WOTPivotNodeProtocol> node = [self.model itemAtIndexPath:indexPath];
        return node == nil ? PivotStickyTypeFloat : node.stickyType;
    }];

    [self.navigationController.navigationBar setDarkStyle];
    
    [self.collectionView registerNib:[UINib nibWithNibName:NSStringFromClass([WOTTankPivotDataCollectionViewCell class]) bundle:nil] forCellWithReuseIdentifier:NSStringFromClass([WOTTankPivotDataCollectionViewCell class])];
    [self.collectionView registerNib:[UINib nibWithNibName:NSStringFromClass([WOTTankPivotFilterCollectionViewCell class]) bundle:nil] forCellWithReuseIdentifier:NSStringFromClass([WOTTankPivotFilterCollectionViewCell class])];
    [self.collectionView registerNib:[UINib nibWithNibName:NSStringFromClass([WOTTankPivotFixedCollectionViewCell class]) bundle:nil] forCellWithReuseIdentifier:NSStringFromClass([WOTTankPivotFixedCollectionViewCell class])];
    [self.collectionView registerNib:[UINib nibWithNibName:NSStringFromClass([WOTTankPivotEmptyCollectionViewCell class]) bundle:nil] forCellWithReuseIdentifier:NSStringFromClass([WOTTankPivotEmptyCollectionViewCell class])];
    
    self.fetchController = [[WOTDataTanksFetchController alloc] initWithNodeFetchRequestCreator:self nodeCreator:self];
    self.model = [[WOTPivotDataModel alloc] initWithFetchController:self.fetchController modelListener:self nodeCreator:self];
    [self.model loadModel];
}

#pragma mark - UICollectionViewDataSource
- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {

    UICollectionViewCell *result = nil;
    
    id<WOTPivotNodeProtocol> node = [self.model itemAtIndexPath:indexPath];
    if ([node isKindOfClass:[WOTPivotRowNode class]]) {
        
        result = [collectionView dequeueReusableCellWithReuseIdentifier:NSStringFromClass([WOTTankPivotFixedCollectionViewCell class]) forIndexPath:indexPath];
        
        WOTTankPivotFixedCollectionViewCell *row = (WOTTankPivotFixedCollectionViewCell *)result;
        row.textValue = node.name;
    }
    
    if ([node isKindOfClass:[WOTPivotColNode class]]) {
        
        result = [collectionView dequeueReusableCellWithReuseIdentifier:NSStringFromClass([WOTTankPivotFixedCollectionViewCell class]) forIndexPath:indexPath];
        
        WOTTankPivotFixedCollectionViewCell *fixed = (WOTTankPivotFixedCollectionViewCell *)result;
        fixed.textValue = node.name;
    }
    
    if ([node isKindOfClass:[WOTPivotFilterNode class]]) {
        
        result = [collectionView dequeueReusableCellWithReuseIdentifier:NSStringFromClass([WOTTankPivotFilterCollectionViewCell class]) forIndexPath:indexPath];
    }
    
    if ([node isKindOfClass:[WOTPivotDataNode class]]) {
        
        result = [collectionView dequeueReusableCellWithReuseIdentifier:NSStringFromClass([WOTTankPivotDataCollectionViewCell class]) forIndexPath:indexPath];
        WOTTankPivotDataCollectionViewCell *dataCell = (WOTTankPivotDataCollectionViewCell *)result;
        dataCell.dataViewColor = node.dataColor;
        
        Tanks *tank = (Tanks *)node.data1;
        dataCell.symbol = node.name;
        dataCell.dpm = [tank.dpm suffixNumber];
        dataCell.mask = tank.invisibility;
        dataCell.visibility = [tank.visionRadius suffixNumber];
    }
    return result;
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {

    return [self.model itemsCountWithSection:section];
}

#pragma mark - UICollectionViewDelegate
- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    id<WOTPivotNodeProtocol> node = [self.model itemAtIndexPath:indexPath];
    if ([node isKindOfClass:[WOTPivotDataNode class]]) {
        Tanks* tank = (Tanks *)node.data1;
        WOTTankModuleTreeViewController *configurationSelector = [[WOTTankModuleTreeViewController alloc] initWithNibName:NSStringFromClass([WOTTankModuleTreeViewController class]) bundle:nil];
        configurationSelector.tankId = tank.tank_id;
        [configurationSelector setCancelBlock:^(){

            [self.navigationController popViewControllerAnimated:YES];
        }];
        [configurationSelector setDoneBlock:^(id configuration){

            [self.navigationController popViewControllerAnimated:YES];
        }];
        [self.navigationController pushViewController:configurationSelector animated:YES];
    }
}

#pragma mark - WOTPivotDataModelListener

- (void)modelDidLoad {
    [self.collectionView reloadData];
}

- (void)modelDidFailLoadWithError:(NSError * _Nonnull)error {

}

- (NSArray *)metadataItems {

    WOTPivotTemplates *templates = [[WOTPivotTemplates alloc] init];

    id<WOTPivotNodeProtocol> levelTier = [templates.vehicleTier asType: PivotMetadataTypeColumn];
    id<WOTPivotNodeProtocol> levelPrem = [templates.vehiclePremium asType: PivotMetadataTypeColumn];
    NSArray *cols = [templates permutateWithPivotNodes:@[levelPrem, levelTier] as:PivotMetadataTypeColumn];

    id<WOTPivotNodeProtocol> levelNati = [templates.vehicleNation asType: PivotMetadataTypeRow];
    id<WOTPivotNodeProtocol> levelType = [templates.vehicleType asType: PivotMetadataTypeRow];
    NSArray *rows = [templates permutateWithPivotNodes:@[levelNati,levelType] as:PivotMetadataTypeRow];

    NSArray *filters = [self pivotFilters];
    NSMutableArray *result = [[NSMutableArray alloc] init];
    [result addObjectsFromArray:cols];
    [result addObjectsFromArray:rows];
    [result addObjectsFromArray:filters];
    return result;
}

- (NSArray *)pivotFilters {
    id<WOTPivotNodeProtocol> node = [[WOTPivotFilterNode alloc] initWithName:@"Filter"];
    return @[node];
}


@end

File: WOT-iOS/WOT-iOS/ui/WOTTankPivotViewController/WOTTankPivotViewController.swift
//
//  WOTTankPivotViewController.swift
//  WOT-iOS
//
//  Created on 8/10/18.
//  Copyright © 2018. All rights reserved.
//

import Foundation
import WOTPivot

typealias WOTTankPivotCompletionCancelBlock = () -> Void
typealias WOTTankPivotCompletionDoneBlock = (_ configuration: Any) -> Void

open class WOTPivotViewController: UIViewController {
    @IBOutlet open var collectionView: UICollectionView?

    @IBOutlet open var flowLayout: WOTPivotLayout? {
        didSet {
            flowLayout?.relativeContentSizeBlock = { [weak self] in
                let size = self?.model.contentSize
                return size ?? CGSize.zero
            }
            flowLayout?.itemRelativeRectCallback = { [weak self] (indexPath) in
                let itemRect = self?.model.itemRect(atIndexPath: indexPath  as NSIndexPath)
                return itemRect ?? CGRect.zero
            }
            flowLayout?.itemLayoutStickyType = { [weak self] (indexPath) in
                let node = self?.model.item(atIndexPath: indexPath as NSIndexPath)
                return node?.stickyType ?? .float
            }
        }
    }

    lazy var refreshControl: WOTPivotRefreshControl = {
        return WOTPivotRefreshControl(target: self, action: #selector(WOTTankPivotViewController.refresh(_:)))
    }()

    open func pivotModel() -> WOTPivotDataModelProtocol {
        return WOTPivotDataModel(enumerator: WOTNodeEnumerator.sharedInstance)
    }

    lazy var model: WOTPivotDataModelProtocol  = {
        return pivotModel()
    }()

    static var openedPopoverKey: UInt8 = 0
    var hasOpenedPopover: Bool {
        get {
            guard let result = objc_getAssociatedObject(self, &WOTTankPivotViewController.openedPopoverKey) as? Bool else {
                return false
            }
            return result
        }
        set {
            objc_setAssociatedObject(self, &WOTTankPivotViewController.openedPopoverKey, newValue, .OBJC_ASSOCIATION_ASSIGN)
        }
    }

    func closePopover() {
        self.hasOpenedPopover = false
    }

    func openPopover() {
        self.hasOpenedPopover = true
    }

    open func cell(forDataNode node: WOTPivotNodeProtocol, at indexPath: IndexPath) -> UICollectionViewCell {
        return UICollectionViewCell()
    }

    open func cell(forNode node: WOTPivotNodeProtocol, at indexPath: IndexPath) -> UICollectionViewCell {
        switch node.cellType {
        case .column, .row: return self.cell(forFixedNode: node, at: indexPath)
        case .filter:       return self.cell(forFilterNode: node, at: indexPath)
        case .data:         return self.cell(forDataNode: node, at: indexPath)
        case .dataGroup:    return self.cell(forDataGroupNode: node, at: indexPath)
        }
    }

    open func cell(forFixedNode node: WOTPivotNodeProtocol, at indexPath: IndexPath) -> UICollectionViewCell {
        return UICollectionViewCell()
    }

    open  func cell(forFilterNode node: WOTPivotNodeProtocol, at indexPath: IndexPath) -> UICollectionViewCell {
        return UICollectionViewCell()
    }

    open func cell(forDataGroupNode node: WOTPivotNodeProtocol, at indexPath: IndexPath) -> UICollectionViewCell {
        return UICollectionViewCell()
    }

    override open func viewDidLoad() {
        super.viewDidLoad()

        self.collectionView?.bounces = true
        self.collectionView?.alwaysBounceHorizontal = false
        self.collectionView?.alwaysBounceVertical = false

        if #available(iOS 10.0, *) {
            self.collectionView?.refreshControl = self.refreshControl
        } else {
            self.collectionView?.addSubview(self.refreshControl)
        }

        self.navigationController?.navigationBar.setDarkStyle()

        self.registerCells()
        self.model.loadModel()
    }

    open func registerCells() {}
}

extension WOTPivotViewController: UICollectionViewDataSource {
    public func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        guard let node = self.model.item(atIndexPath: indexPath as NSIndexPath) else {
            return UICollectionViewCell()
        }
        return self.cell(forNode: node, at: indexPath)
    }

    public func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return self.model.itemsCount(section: section)
    }
}

extension WOTPivotViewController: UICollectionViewDelegate {
    public func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        guard let pivotNode = self.model.item(atIndexPath: indexPath as NSIndexPath)  else {
            return
        }

        guard self.hasOpenedPopover == false else {
            self.closePopover()
            return
        }

        let fetchedObject = pivotNode.data1 as? NSManagedObject
        switch pivotNode.cellType {
        case .data: openTankDetail(data: fetchedObject)
        case .dataGroup: openPopover()
        default: break
        }
    }

    private func openTankDetail(data: NSManagedObject?) {
        guard let vehicle = data as? Vehicles else {
            return
        }
        let config = WOTTankModuleTreeViewController(nibName: String(describing: WOTTankModuleTreeViewController.self), bundle: nil)
        config.tank_Id = vehicle.tank_id
        config.cancelBlock = {
            self.navigationController?.popViewController(animated: true)
        }
        config.doneBlock = { (conf) in
            self.navigationController?.popViewController(animated: true)
        }
        self.navigationController?.pushViewController(config, animated: true)
    }
}

@objc(WOTTankPivotViewController)
class WOTTankPivotViewController: WOTPivotViewController {
    var cancelBlock: WOTTankPivotCompletionCancelBlock?
    var doneBlock: WOTTankPivotCompletionDoneBlock?
    var fetchedResultController: NSFetchedResultsController<NSFetchRequestResult>?

    override func pivotModel() -> WOTPivotDataModelProtocol {
        return WOTTankPivotModel(modelListener: self, dataProvider: WOTPivotAppManager.sharedInstance.coreDataProvider)
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        let items = [UIBarButtonItem(barButtonSystemItem: UIBarButtonItem.SystemItem.action, target: self, action: #selector(WOTTankPivotViewController.openConstructor(_:)))]
        self.navigationItem.setRightBarButtonItems(items, animated: false)
    }

    @objc
    func openConstructor(_ sender: Any) {
        let vc = WOTPivotConstructorViewController(nibName: "WOTPivotConstructorViewController", bundle: Bundle.main)
        self.navigationController?.pushViewController(vc, animated: true)
    }

    @objc
    func refresh(_ refreshControl: UIRefreshControl) {
        self.model.loadModel()
    }

    static var registeredCells: [UICollectionViewCell.Type] = {
        return [WOTTankPivotDataCollectionViewCell.self,
                WOTTankPivotFilterCollectionViewCell.self,
                WOTTankPivotFixedCollectionViewCell.self,
                WOTTankPivotEmptyCollectionViewCell.self,
                WOTTankPivotDataGroupCollectionViewCell.self]
    }()

    @objc
    override public func registerCells() {
        WOTTankPivotViewController.registeredCells.forEach { (type) in
            let clazzStr = String(describing: type)
            let nib = UINib(nibName: clazzStr, bundle: nil)
            self.collectionView?.register(nib, forCellWithReuseIdentifier: clazzStr)
        }
    }

    override func cell(forDataNode node: WOTPivotNodeProtocol, at indexPath: IndexPath) -> UICollectionViewCell {
        let ident = String(describing: WOTTankPivotDataCollectionViewCell.self)
        guard let cell = self.collectionView?.dequeueReusableCell(withReuseIdentifier: ident, for: indexPath) as? WOTTankPivotDataCollectionViewCell else {
            return UICollectionViewCell()
        }
        cell.symbol = node.name
        cell.dataViewColor = node.dataColor
        cell.uuid = node.index
        return cell
    }

    override func cell(forDataGroupNode node: WOTPivotNodeProtocol, at indexPath: IndexPath) -> UICollectionViewCell {
        let ident = String(describing: WOTTankPivotDataGroupCollectionViewCell.self)
        guard let cell = self.collectionView?.dequeueReusableCell(withReuseIdentifier: ident, for: indexPath) as? WOTTankPivotDataGroupCollectionViewCell else {
            return UICollectionViewCell()
        }
        return cell
    }

    override public func cell(forFixedNode node: WOTPivotNodeProtocol, at indexPath: IndexPath) -> UICollectionViewCell {
        let ident = String(describing: WOTTankPivotFixedCollectionViewCell.self)
        guard let result = self.collectionView?.dequeueReusableCell(withReuseIdentifier: ident, for: indexPath) as?  WOTTankPivotFixedCollectionViewCell else {
            return UICollectionViewCell()
        }
        result.textValue = node.name
        return result
    }

    override public func cell(forFilterNode node: WOTPivotNodeProtocol, at indexPath: IndexPath) -> UICollectionViewCell {
        let ident = String(describing: WOTTankPivotFilterCollectionViewCell.self)
        guard let result = self.collectionView?.dequeueReusableCell(withReuseIdentifier: ident, for: indexPath) else {
            return UICollectionViewCell()
        }
        return result
    }
}

extension WOTTankPivotViewController {
    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        self.refreshControl.contentOffset = scrollView.contentOffset
    }
}

extension WOTTankPivotViewController: WOTDataModelListener {
    func modelDidLoad() {
        self.collectionView?.reloadData()
        self.refreshControl.endRefreshing()
    }

    func modelDidFailLoad(error: Error) {
        self.refreshControl.endRefreshing()
    }
}
